### PR TITLE
Added onnx config whisper

### PR DESCRIPTION
**Pull request by mht-sharma**
# What does this PR do?

Fixes # (issue)

This PR adds onnx config and helper functions for export to onnx via optimum and transformers.onnx

### CODE REVIEW

File: src/transformers/models/whisper/configuration_whisper.py
Comment (lewtun) [inline src/transformers/models/whisper/configuration_whisper.py:286]: nit: in `transformers` we prefer being explicit with variable names
```suggestion
        num_past_key_values_per_layer = 4
```

File: src/transformers/models/whisper/configuration_whisper.py
Comment (lewtun) [inline src/transformers/models/whisper/configuration_whisper.py:290]:
```suggestion
        for i in range(num_decoder_layers * num_past_key_values_per_layer):
```

File: src/transformers/models/whisper/configuration_whisper.py
Comment (lewtun) [inline src/transformers/models/whisper/configuration_whisper.py:299]: should the return type be this?

```suggestion
    def get_encoder_config(self, encoder_config: PretrainedConfig) -> WhisperEncoderOnnxConfig:
```

File: src/transformers/models/whisper/configuration_whisper.py
Comment (lewtun) [inline src/transformers/models/whisper/configuration_whisper.py:327]: nit:
```suggestion
            feature (`str`, *optional*, defaults to `default`):
```

File: src/transformers/models/whisper/configuration_whisper.py
Comment (lewtun) [inline src/transformers/models/whisper/configuration_whisper.py:329]: nit:
```suggestion
            use_past (bool, *optional*, defaults to `False`):
```

**Review by lewtun — CHANGES_REQUESTED** [review]
Thanks for working on this @mht-sharma - it's looking really good 🗣!!

I left a few nits and a general question about whether we should refactor the dummy data generation into a separate function.

Let's also add the logic to enable the ONNX export to work with `transformers.onnx` so we can add the model to the unit tests

File: src/transformers/models/whisper/configuration_whisper.py
Comment (echarlaix) [inline src/transformers/models/whisper/configuration_whisper.py:241]: This config handling the export of encoder/decoder model have already been integrated to [optimum](https://github.com/huggingface/optimum/blob/v1.4.0/optimum/onnx/configuration.py#L282). Given the current efforts to migrate the ONNX export from transformers to optimum, I don't think it makes sense to duplicate this classes in transformers (given that it will be soon depreciated). Why not move this PR to optimum ? @mht-sharma  @lewtun @michaelbenayoun

Comment (lewtun) [reply src/transformers/models/whisper/configuration_whisper.py:240]: Good point @echarlaix - I originally thought this PR was needed to enable the `from_transformers=True` feature in `optimum`. I now agree this would be best moved to `optimum` as part of https://github.com/huggingface/optimum/pull/420

WDYT @mht-sharma? Shall we close this PR and move the logic over?

Comment (mht-sharma) [reply src/transformers/models/whisper/configuration_whisper.py:240]: Hi @lewtun @echarlaix we may not be able to close this PR completely. Optimum still uses some of the components from the transformers.onnx package. Hence, to support the above model we would need to create placeholders for WhisperOnnxConfig, support for speech modality, AutoClassForSpeechSeq2Seq, audio data generation etc in the transformers repository.

For example
https://github.com/huggingface/optimum/blob/main/optimum/onnxruntime/modeling_seq2seq.py#L433

```python
        model = FeaturesManager.get_model_from_feature(cls.export_feature, model_id)
        _, model_onnx_config = FeaturesManager.check_supported_model_or_raise(model, feature=cls.export_feature)
        onnx_config = model_onnx_config(model.config)
```
The above line requires that the model is supported in transformers.onnx also.

Best we could do is to copy the Whisper ONNX configs in optimum also and use for optimum pipelines.

Comment (lewtun) [reply src/transformers/models/whisper/configuration_whisper.py:240]: Thanks for clarifying the need for this PR @mht-sharma - I overlooked this dependency in `optimum` 🙈

In the interest of moving fast + bringing value to the community, it probably makes sense to indeed continue with this PR in `transformers` since the ONNX migration to `optimum` will probably take some more time. WDYT @echarlaix ?

Comment (echarlaix) [reply src/transformers/models/whisper/configuration_whisper.py:240]: We need an `WhisperOnnxConfig` such as [`T5OnnxConfig`](https://github.com/huggingface/transformers/blob/v4.23.1/src/transformers/models/t5/configuration_t5.py#L143) which would keep the export behavior of whisper similar to other encoder/decoder model of `transformers.onnx` (export of entire model in one shot) but keep the `EncoderOnnxConfig` / `DecoderOnnxConfig` in `optimum`

File: src/transformers/onnx/config.py
Comment (mht-sharma) [reply src/transformers/onnx/config.py]: Added a function to generate dummy audio with custom length and sampling rates

File: src/transformers/models/whisper/configuration_whisper.py
Comment (echarlaix) [reply src/transformers/models/whisper/configuration_whisper.py:244]: What do you think @mht-sharma @lewtun ?

Comment (lewtun) [reply src/transformers/models/whisper/configuration_whisper.py:244]: OK, if we can enable Whisper to run on ONNX Runtime in `optimum` without any dependency on internal `transformers` logic, then I think @echarlaix's proposal to migrate everything there will help us be future proof. It will also make the final PR a bit easier to test :)

Apologies if I caused confusion by flipping back and forth - this will become much simpler once we have the ONNX exporter in `optimum`!

Comment (mht-sharma) [reply src/transformers/models/whisper/configuration_whisper.py:244]: Thanks @lewtun @echarlaix  I will migrate the changes.

File: src/transformers/onnx/config.py
Comment (lewtun) [inline src/transformers/onnx/config.py:386]: nit:
```suggestion
        Generate inputs for ONNX Runtime using the reference model inputs.
```

File: src/transformers/onnx/config.py
Comment (lewtun) [inline src/transformers/onnx/config.py:389]: nit:
```suggestion
            reference_model_inputs ([`Mapping[str, Tensor]`):
```

File: src/transformers/onnx/config.py
Comment (lewtun) [inline src/transformers/onnx/config.py:393]: nit:
```suggestion
            `Mapping[str, Tensor]`: The mapping holding the kwargs to provide to the model's forward function
```

File: src/transformers/onnx/config.py
Comment (lewtun) [inline src/transformers/onnx/config.py:384]: If I'm not mistaken, this function doesn't do anything - why do we need it?

**Review by lewtun — COMMENTED** [review]
Thanks for iterating @mht-sharma - this looks very close to being ready 🔥 !!

I've left some nits and a question, but otherwise this all looks good :)

File: src/transformers/onnx/config.py
Comment (mht-sharma) [reply src/transformers/onnx/config.py:384]: Hi @lewtun , this is true that this function does not do anything for existing models. However, this function can be overridden in some cases where the model inputs and ONNX Runtime inputs are different. This was needed when exporting the decoder in encoder-decoder models using encoder_outputs. For example: Check the following Decoder config in optimum where I am using the function [DecoderOnnxConfig](https://github.com/huggingface/optimum/pull/420/files#diff-c27ea812737bc6ccfe34f92a4ff0d1ec473a41b8c8012bfdb08bb22a46104ddeR342). 

Since we are waiting for the optimum PR to merge and migrating the changes this is no longer needed in the current merge.

But we still need to update the `VisionEncoderDecoderConfig` using encoder_outputs in that case we would need such function. Should we merge with this and utilise this in new `VisionEncoderDecoder` PR (In case the user wants to create the PR for this change, this option would become easier for him)? Or we could remove it from here and add the function along with the new PR?

Comment (lewtun) [reply src/transformers/onnx/config.py:384]: I see, thanks for the clarification! Looking at the `optimum` code, it seems like we use this function to add new fields to the ORT inputs - is there a reason we can't capture that logic in a single `generate_dummy_inputs()` function associated with the ONNX config?

(I'm happy to include this function as is, just trying to understand if we really need it or not)

Comment (mht-sharma) [reply src/transformers/onnx/config.py:384]: The function (at least in this case) update the existing keys in the input dict, but the values remains the same.

For exporting these models we would require 2 different input sets. This is because the model.forward has a different input signature (since this is the full model before export) that the generated ONNX model (only decoder is exported). Therefore we need a way to alter the existing model inputs to run inference with both models in validate_model_outputs.

However, I think creating a separate function is a much cleaner way. But I am open to any suggestions. We can add the logic in the `generate_dummy_inputs` by adding a new argument to the function maybe called `ort_inputs=True/False` and for these models we return different sets of inputs. But this will require updation of all the OnnxConfigs having this function.

Comment (lewtun) [reply src/transformers/onnx/config.py:384]: OK great, now I understand well why we need this - thanks. IMO it's fine to include this function in this PR if we add a small note in the docstring like "Override to run inference with seq2seq models which have the encoder and decoder exported as separate ONNX files."

Comment (mht-sharma) [reply src/transformers/onnx/config.py:384]: Done

File: src/transformers/onnx/config.py
Comment (mht-sharma) [reply src/transformers/onnx/config.py]: Done

File: src/transformers/onnx/config.py
Comment (mht-sharma) [reply src/transformers/onnx/config.py]: Done

**Review by lewtun — APPROVED** [review]
Thanks for the final round of iterations @mht-sharma - this now LGTM!

Let's wait for final approval from @sgugger before merging

File: src/transformers/models/whisper/configuration_whisper.py
Comment (sgugger) [inline src/transformers/models/whisper/configuration_whisper.py:26]: Let's put the right module here :-)

File: src/transformers/onnx/config.py
Comment (sgugger) [inline src/transformers/onnx/config.py:319]:
```suggestion
            time_duration (`float`, *optional* defaults to 5.0)
```
Let's be consistent with the signature!

File: src/transformers/onnx/config.py
Comment (sgugger) [inline src/transformers/onnx/config.py:351]:
```suggestion
            input_token = preprocessor.unk_token if preprocessor.unk_token is not None else "0"
```
We don't rely on Python bool magic conversion in the library, so let's test explicitly.

**Review by sgugger — APPROVED** [review]
Thanks for adding this! There a few details to fix, but then we should be good to merge.

File: src/transformers/onnx/config.py
Comment (mht-sharma) [reply src/transformers/onnx/config.py]: Done

File: src/transformers/onnx/config.py
Comment (mht-sharma) [reply src/transformers/onnx/config.py]: Done, Added explicit tests for None and empty string

File: src/transformers/models/whisper/configuration_whisper.py
Comment (mht-sharma) [reply src/transformers/models/whisper/configuration_whisper.py]: Updated

**Comment by lewtun** [comment]
Thanks for the review @sgugger and catching those last issues - I've checked the latest changes and think this can now be merged @mht-sharma

**Comment by zara0m** [comment]
Hello,

I tried to generate onnx model using [docs](https://huggingface.co/docs/transformers/serialization#configuration-based-approach),
To inference, I passed audio features and decoder_input_id , and the output was two array with the shape of (1, 2, 768), (1, 1500, 768). Could you please help me how should I use these outputs for generating transcription? 

Thank you.

**Comment by mht-sharma** [comment]
Hi @zara0m please follow the example in this PR for export and  inference using ONNX model. https://github.com/huggingface/optimum/pull/420

**Comment by zara0m** [comment]
> Hi @zara0m please follow the example in this PR for export and inference using ONNX model. [huggingface/optimum#420](https://github.com/huggingface/optimum/pull/420)

Thank you very much for your help and quick response!

I tested it with base and small model for some non-english audios, but the outputs were not similar to whisper model, or maybe it translates instead of transcribing, how can I fix this?
Also, is there any way that I can have begin/end time of each sentence? (like the transcribe function of whisper model)

Thank you.

**Pull request merged**
